### ISSUE TITLE
[spec] implement support of OpenID Connect (backend)

### ISSUE THREAD

**Issue by mhuin** [change-metrics/monocle#877]
This spec aims to add support for authentication via OpenID Connect in Monocle. This mode of authentication will be optional, meaning that if not set in the application's configuration, the legacy authentication workflow will be enforced by default. Both authentication modes will be mutually exclusive. This spec complements [ADR 12: OpenID Connect Authentication](https://github.com/change-metrics/monocle/blob/master/doc/adr/0012-OpenID-connect-authentication.md) and is a follow-up for [implement support of OpenID Connect (web UI)](https://github.com/change-metrics/monocle/issues/874).

# Current state

The search endpoint accepts a "username" parameter. The backend uses this parameter in search queries with the "self" keyword. No control or verification is done on the username, so anyone can "impersonate" any given user.

Example cURL query:

`curl 'http://monocle.usersys.redhat.com/api/2/search/query' -X POST --data-raw '{"index":"openstack","username":"XYZ","query":"from:now-3weeks author:self","query_type":"QUERY_CHANGES_LIFECYCLE_STATS","order":{"field":"updated_at","direction":"DESC"},"limit":0,"change_id":""}'`

# Proposed change

Add a backend configuration setting for specifying the OpenID Connect authority (ie the well-known endpoint), the expected audience, and the expected issuer. These settings are required for token validation.

It is required that the Monocle backend service can send HTTP queries to the OIDC authority, to fetch the public keys needed for token verification.

Add support for OpenID Connect JWTs in the API. The JWT is passed as the "Authorization" header
to the API. It is implied that if the OpenID Connect configuration section is set in the backend, the following workflow must be applied:

1. If no authorization header is provided, reject the query with a HTTP 401 error (Unauthorized).
2. If a authorization header is provided, the token goes through basic validation: signature verification (requires HTTP calls to the OIDC authority), audience check, issuer check, expiry time check. If any of these checks fails, reject the query with a HTTP 401 error (Unauthorized).
3. if the `username` field is present in the data, it must match the `preferred_username` in the JWT. If this check fails, reject the query with a HTTP 403 error (Forbidden). No check is performed if the field isn't present; the backend will directly use the `preferred_username` claim from the JWT to match the `self` query keyword.
4. if all checks passed, return the query results.

## Magic Token

For testing and cli users, we need to be able to easily generate tokens. We will create a utility script that will generate a JWT from a special issuer (_MAGIC_ ?). The script will require a key pair for signing and verifying keys, so the script could also be used to generate these.

# Impact

- A new haskell dependency will be added to the project to handle JWT verification.
- We must ensure POST endpoints follow the CORS standards (preflight exchange, headers validation and responses).
- cURL users need to figure out a way to get a valid JWT to use the protected endpoint. This can be documented but requires specific configuration on the Identity Provider and is probably out of scope for this spec.

# Follow-ups

Once token verification is implemented, an authorizations system based on token claims can be added on top of this. For example, only users belonging to the "managers" group could be allowed to run queries without the "self" keyword.

The API could be modified to drop the "username" parameter, and only get user info from the JWT. In the case of an unauthenticated Monocle instance, the web UI could be modified to match "self" in user inputs to the self-provided username prior to sending the query.

# 06/07/2022 - Amending

After some testing we've figured out that the *OIDC Implicit flow* implemented via the Monocle WEB App will lack of flexibility and probably limit the integration of Monocle with various OIDC OP. Then it has been decided to rely on the backend part to handle the *OIDC Authorization code flow*.

Tasks:

- [x] Read Authorization header on each endpoint
- [x] Implement a function to validate a JWT token
- [x] Add configuration of the OIDC Provider through ENV vars
- [x] Implement a JWT validation test
- [x] Add support for "magic token" : creation, validation, documentation
- [x] All protected endpoints must return a 403 when JWT cannot be validated
- [x] Local JWT must resolve the muid from the Idents DB by looking at aliases in each index
- [x] Search endpoint must use the muid of the related index to resolve the username attribute
- [x] Local JWT issuer must be initialized with a private RSA key (https://hackage.haskell.org/package/jose-0.9/docs/Crypto-JOSE-JWK.html#g:3)
- [x] Add documentation in README.md
- [x] Add changelog entry

**Comment by TristanCacqueray** [comment]
Thanks for the spec! It looks great already, I am looking forward the propose authorization systems :)

Here are some thoughts.

- Instead of saying that the Monocle backend is required to send HTTP queries to the OIDC authority, could we describe the workflow with a sequence diagram?

```plantuml
note left    : on startup
Api -> IDP   : fetch the public key'
...
User -> Api  : send request
Api -> User  : send result if the request contains a valid token, otherwise reject the query with 401
```

- Why should the web UI auth needs to be finalized first? The test will demonstrates the new feature without the web UI. Thus the order should not matter. I would actually prefer that we implement the API first.

- In the impact section, adding a new dependencies is not very important, but I think what matters is that all the endpoints will have to perform token validations.

- If we do the test right, it should be easy to expose the token generation feature for curl users.

- In the follow-ups, we should mention that the crawler api token system could be replaced by the new system.

**Comment by mhuin** [comment]
>     * Why should the web UI auth needs to be finalized first? The test will demonstrates the new feature without the web UI. Thus the order should not matter. I would actually prefer that we implement the API first.

Sure, it could even be done in parallel, but if we want to provide a user-facing change ASAP we should start with the web UI. The API support for OIDC won't be used until it is implemented in the UI; whereas adding OIDC support in the GUI will be usable immediately to support fetching user info, if a provider is available.

>     * In the impact section, adding a new dependencies is not very important, but I think what matters is that all the endpoints will have to perform token validations.

Should all the endpoints support tokens, ie authN/authZ?

>     * If we do the test right, it should be easy to expose the token generation feature for curl users.

Generating a token to be used with cURL is out of scope for Monocle. The usual way to do that is to configure the Identity Provider to accept "Direct Access Grants", ie allow a client to send a username and password directly to the /token endpoint rather than require to do so from the IdP's front end. This goes against best practices for OIDC (the client should never be aware of a user's secrets) so it won't always be available to users.

If this is a problem, we can add to the spec the creation of a utility to generate "magic tokens" that can be handed to users and used in the CLI.

>     * In the follow-ups, we should mention that the crawler api token system could be replaced by the new system.

Shouldn't the API token be long-lived? OIDC JWT are usually short-lived, with the possibility to renew them.

**Comment by TristanCacqueray** [comment]
The specification should not be affected by the priority, or the need for a change ASAP, it's just a specification. Therefor I don't think the propose changed should be done after the web UI, and we should remove the `Once the web UI auth is finalized` part.

I reckon the goal is to prevent anonymous access, thus I think when OpenID is configured, then every route should enforce authentication. And it seems like we should implement "magic tokens" so that the feature can be tested, and I think this will be useful for crawler and further service like companion.

**Comment by morucci** [comment]
This specification is looking good.

I agree that the order does not really matter and both could be implemented in parallel.

The implementation backend side must come with unit tests where the validation of a fake id_token is performed.

Regarding cURL access to the backend API, I'm wondering whether we should provide an API endpoint to allow an authenticated user to generate a token that can be use to request the API w/o a JWT token (backend registers user' tokens in the database).

**Comment by morucci** [comment]
Let's agree on a list of actionable tasks:
- [ ] Read Authorization header on each endpoint
- [ ] Implement a function to validate a JWT token
- [ ] Add configuration section for authentication in dhall-monocle
- [ ] Implement a JWT validation test
